flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/aggregation/PreAggregationOperator.java

```diff
@@ -0,0 +1,249 @@
+ * <ol>
+ *     <li>{@link MergingWindowAssigner} is not supported.</li>
+ *     <li>{@link PreAggregationOperator} emits all of its data on each received watermark.</li>
+ *     <li>On restoring from checkpoint keys can be randomly shuffled between {@link PreAggregationOperator}
+ *     instances</li>.
+ * </ol>
+ */
+@PublicEvolving
+public class PreAggregationOperator<K, IN, ACC, W extends Window>
+	extends AbstractStreamOperator<Tuple3<K, W, ACC>>
+	implements OneInputStreamOperator<IN, Tuple3<K, W, ACC>>, Serializable {
+
+	protected final AggregateFunction<IN, ACC, ?> aggregateFunction;
+	protected final KeySelector<IN, K> keySelector;
+	protected final WindowAssigner<? super IN, W> windowAssigner;
+	protected final TypeInformation<K> keyTypeInformation;
+	protected final TypeInformation<ACC> accumulatorTypeInformation;
+	protected final boolean flushAllOnWatermark;
+	protected final Map<Tuple2<K, W>, ACC> aggregates = new HashMap<>();
+
+	protected transient WindowAssigner.WindowAssignerContext windowAssignerContext;
+	protected transient ListState<Tuple3<K, W, ACC>> aggregatesState;
+
+	/**
+	 * Creates {@link PreAggregationOperator}.
+	 *
+	 * @param aggregateFunction function used for aggregation. Note, {@link AggregateFunction#getResult(Object)} will
+	 *                          not be used.
+	 * @param keySelector
+	 * @param keyTypeInformation
+	 * @param accumulatorTypeInformation
+	 * @param windowAssigner
+	 * @param flushAllOnWatermark flag to control whether all elements should be emitted on any watermark. Check more
+	 *                            information in {@link PreAggregationOperator}.
+	 */
+	public PreAggregationOperator(
+			AggregateFunction<IN, ACC, ?> aggregateFunction,
+			KeySelector<IN, K> keySelector,
+			TypeInformation<K> keyTypeInformation,
+			TypeInformation<ACC> accumulatorTypeInformation,
+			WindowAssigner<? super IN, W> windowAssigner,
+			boolean flushAllOnWatermark) {
+		this.aggregateFunction = checkNotNull(aggregateFunction, "aggregateFunction is null");
+		this.keySelector = checkNotNull(keySelector, "keySelector is null");
+		this.windowAssigner = checkNotNull(windowAssigner, "windowAssigner is null");
+		this.keyTypeInformation = checkNotNull(keyTypeInformation, "keyTypeInformation is null");
+		this.accumulatorTypeInformation = checkNotNull(accumulatorTypeInformation, "accumulatorTypeInformation is null");
+		this.flushAllOnWatermark = flushAllOnWatermark;
+
+		checkNotNull(keyTypeInformation, "keyTypeInformation is null");
+		checkNotNull(accumulatorTypeInformation, "accumulatorTypeInformation is null");
+
+		checkArgument(!(windowAssigner instanceof MergingWindowAssigner),
+			"MergingWindowAssigner is not supported by the PreAggregationOperator");
+	}
+
+	@Override
+	public void open() throws Exception {
+		windowAssignerContext = new WindowAssigner.WindowAssignerContext() {
+			@Override
+			public long getCurrentProcessingTime() {
+				return System.currentTimeMillis();
```

Comment:
This should use the `ProcessingTimeService` that can be retrieve via `getProcessingTimeService()`. Then you can also test the operator with processing time because the test harness allows advancing processing time manually.

Reply:
Changed to throwing `UnsupportedException`, since processing time doesn't make sense combined with pre aggregation.